lib/galaxy/webapps/galaxy/buildapp.py

```diff
@@ -736,6 +736,12 @@ def connect_invocation_endpoint(endpoint_name, endpoint_suffix, action, conditio
                           action='api_key',
                           conditions=dict(method=["POST"]))
 
+    webapp.mapper.connect('api_key',
+                          '/api/users/{id}/api_key',
+                          controller='users',
+                          action='get_or_create_api_key',
+                          conditions=dict(method=["GET"]))
```

Comment:
We've got separate endpoints for get_api_key (GET to /api/users/{id}/api_key/inputs) and set_api_key (POST to /api/users/{id}/api_key'), this feels like it's muddying the waters. Maybe we should just re-route /api/users/{id}/api_key/inputs to `GET /api/users/{id}/api_key/` ?

Reply:
The response of `/inputs` is a mess. I would much prefer only a json encoded string be returned like `/api/authenticate/baseauth`

Reply:
Agreed - those endpoints are gross to work with because they're designed for the form API. I like the idea of an API for doing this that is designed to work as an API.

Reply:
The other issue is unless you expose the "get or create" function, it creates a race condition where the API key is overwritten by a second competing POST request generating an API key.

Reply:
I mean, a GET shouldn't modify the underlying resource, and this does that. If you only see GET and `'/api/users/{id}/api_key'` you'd never think it could create an API key. At least make it a POST, please ?

Reply:
I think GET is actually the most reasonable thing here, actually -- it's still an idempotent operation.  PUT seems weird since you're not actually pushing any data.  Let me think about this.

Reply:
POST is the correct operation for this, PUT wouldn't be.

Reply:
Care to elaborate?  POST doesn't imply idempotency, which I see as the most relevant attribute of a 'get or create'.  I still think GET is the right action here.

Reply:
https://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html#sec9

I suppose it is a practice vs specification dilemma. You are right, according to the spec a GET operation would be more appropriate. I think the larger issue is that the spec doesn't address whether GET should or shouldn't have side-effects. POST is supposed to create a subordinate at the specified path, which makes it inappropriate in that regard. PUT can't be used because it is supposed to have a request body that is assigned to the resource.

Because we are requesting the resource specified by the URI and not a subordinate, GET should be used.

Reply:
Alright, GET it is.